### PR TITLE
Restructure Keras Scikit-Learn wrappers to better implement Scikit-Learn API

### PR DESCRIPTION
**Pull request by adriangb**
This is a modification of #32533. I am opening a new PR because that one seems stalled and I made a lot of changes/improvements (but keeping the same idea).

A quick summary:
The existing `scikit-learn` wrappers for Keras models are not compatible with many `scikit-learn` functions. Additionally, they require that dataset dimensions be determined _before_ calling the `fit` method, which is unlike the `scikit-learn` estimators and makes it hard to build dynamically adaptable pipelines.

What this PR does:
This PR does _not_ change any API.
By moving the storage of parameters from `self.sk_params` to `self.__dict__`, compatibility with a lot of the `scikit-learn` functionalities are improved. Additionally, I gave the model building function the ability to request the data that will be fitted (to determine dimensions) as well as any other attributes of the wrapper instance. Finally, I enabled copying/pickling of wrapped models as well as the ability to wrap instances of `Model`, which should allow for greater flexibility in incorporating into an exsiting `Keras` workflow.

How is it tested:
All existing tests are working and were unchanged. This confirms that there were no API changes. New tests were added for all of the new functionality as well as some of the most common `scikit-learn` operations that were previously broken.
I would like to credit @daviddiazvico , the author of the original PR: I borrowed a lot of the tests he had written as well as the original idea for fixing these issues. I will make him a co-author of the final commit if this gets approved.

### CODE REVIEW

**Comment by daviddiazvico** [comment]
Thanks for your effort @adriangb ! I'm sorry I couldn't write back before.

Of course, I don't have any problem with you updating the other PR, but I'm not sure if I have to do something to allow you to edit it. I've added you as a collaborator to my tf fork.

In any case, this is a nice improvement for tf.keras in my opinion, but I don't know if the development team is interested. My [PR](https://github.com/tensorflow/tensorflow/pull/32533) has been open for a while and [the same changes](https://github.com/keras-team/keras/pull/12833) were proposed in the keras repo about a year ago now.

**Comment by adriangb** [comment]
I'm hoping that maybe the reviewers were just a bit busy?

I feel like the wrappers are used quite often, especially in beginner tutorial. If we want more users for tf/keras, I think it's important that the initial experience be seamless.

In addition to your PR, this resolves several issues: #33204, #36074, #34689 and #36137. There are/will be more, like the [comment](https://github.com/tensorflow/tensorflow/pull/32533#issuecomment-590993294) I posted on your PR.

**Comment by adriangb** [comment]
I'm hoping that maybe the reviewers were just a bit busy?

I feel like the wrappers are used quite often, especially in beginner tutorials. I think it's important that the initial experience be seamless.

In addition to your PR, this resolves several issues: #33204, #36074, #34689 and #36137. There are/will be more, like the [comment](url) I posted on your PR.

I'm hoping we can at least get some feedback from @fchollet or @pavithrasv regarding interest in these changes.

File: tensorflow/python/keras/wrappers/scikit_learn.py
Comment (adriangb) [inline tensorflow/python/keras/wrappers/scikit_learn.py]: This should be doable without `h5py`. I'm not sure what the prefered format to use is, or if it is better to use `serialize_keras_object` or something like that.

File: tensorflow/python/keras/wrappers/scikit_learn.py
Comment (adriangb) [inline tensorflow/python/keras/wrappers/scikit_learn.py]: I realized that `self.__class__.__mro__` should be `reversed(self.__class__.__mro__)`. I will change in a future commit.

File: tensorflow/python/keras/wrappers/scikit_learn.py
Comment (adriangb) [reply tensorflow/python/keras/wrappers/scikit_learn.py]: I take this back: the entirety of Tensorflow/Keras is hardcoded to filenames or `h5py.File` objects. 
 
An aside, unrelated to this PR: It is a real shame that we can't save to arbitrary file streams. I understand that HDF5 can have advantages over a more straightforward compress and dump to file stream method, but if it is the _only_ way to save to an io stream, it should be made a requirement, not a sort of optional dependancy. Maybe I'll submit a seperate PR to enable at least `io.BytesIO`.

Comment (adriangb) [reply tensorflow/python/keras/wrappers/scikit_learn.py]: @qlzh727 I know you said in the other PR that you don't really work with the sklearn wrappers, but any input on this more general issue of how to serialize Keras models to memory in a picklable form?

Comment (adriangb) [reply tensorflow/python/keras/wrappers/scikit_learn.py]: I came up with the following. It _seems_ to work for all of the types of models I had, but maybe it fails for custom models?
```python3
from tensorflow.python.keras.saving import saving_utils


SavedKerasModel = namedtuple(
    "SavedKerasModel", "cls model_config training_config weights",
)
```


```python3
def __pack_obj(obj):
    try:
        return copy.deepcopy(obj)
    except TypeError:
        if isinstance(obj, Model):
            model_metadata = saving_utils.model_metadata(obj)
            training_config = model_metadata["training_config"]
            model_config = obj.get_config()
            weights = obj.get_weights()
            return SavedKerasModel(
                cls=obj.__class__,
                model_config=model_config,
                weights=weights,
                training_config=training_config,
            )
```
```python3
def __unpack_obj(obj):
    if isinstance(obj, SavedKerasModel):
        restored_model = obj.cls.from_config(obj.model_config)
        training_config = obj.training_config
        restored_model.compile(
            **saving_utils.compile_args_from_training_config(training_config)
        )
        restored_model.set_weights(obj.weights)
        return restored_model
```

Comment (qlzh727) [reply tensorflow/python/keras/wrappers/scikit_learn.py]: @k-w-w who is the expert for save model.

**Comment by adriangb** [comment]
I was able to add built-in support for all of the multi-output modes that Scikit-Learn supports, as well as a framework to easily support multi-input models. This means this would close #34689 as well.

Because of the great modularity of the Functional API, only a limited number of multi-output cases can be automatically supported (those that scikit-learn itself supports) and no multi-input models can be supported out of the box. If a user wants to use a more complex model with the wrappers, they would have to subclass and manually define the desired behavior.

Although the original problem statement ('fix compatibility') was quite large, I do feel that this PR has grown very large. I would personally prefer to split it into smaller PRs (even if that means more work for me), but I will leave that up to the reviewers.

File: tensorflow/python/keras/wrappers/scikit_learn.py
Comment (adriangb) [reply tensorflow/python/keras/wrappers/scikit_learn.py]: Any updates @qlzh727 @k-w-w ?

Comment (k-w-w) [reply tensorflow/python/keras/wrappers/scikit_learn.py]: This wrapper looks pretty good and should work for non-subclassed models. To support subclassed models, you should save both the class name and config (the results from [generic_utils.serialize_keras_object](https://github.com/tensorflow/tensorflow/blob/v2.1.0/tensorflow/python/keras/utils/generic_utils.py#L191)). 

Calling [`keras.layers.deserialize`](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/serialization.py#L71) will load subclassed models if they were registered with [`tf.keras.utils.generic_utils.register_keras_serializable`](https://www.tensorflow.org/api_docs/python/tf/keras/utils/register_keras_serializable).

Comment (adriangb) [reply tensorflow/python/keras/wrappers/scikit_learn.py]: Thank you @k-w-w. I will look into implementing that. Any chance you can do a quick approval, just so all of the tests can run?

**Comment by adriangb** [comment]
I realized that we actually need to re-implement not only an R^2 score, but also a classifier accuracy score: Keras does not use the `sample_weight` parameter for metrics, but Scikit-Learn does use it for scoring. That and the fact that Keras metrics don't support the Scikit-Learn style multi-output concept, so all of those scores are already being handled manually.

This prompted me to think:  does it make sense to make scikit-learn an optional dependency that is only imported within this module? I can see pros and cons, just playing devil's advocate here.

**Review by k-w-w — DISMISSED** [review]
Approving to run the presubmits.

Can you add tests to make sure that saving and loading works with the wrapper?

**Comment by adriangb** [comment]
> Approving to run the presubmits.
> 
> Can you add tests to make sure that saving and loading works with the wrapper?

Thank you for kicking off those tests!

There are several tests for pickling/unpickling of Functional API models with and without Callbacks, etc. I think the only thing that is missing is a test for subclassed models. I'll add that in the next few days.

**Comment by adriangb** [comment]
Quite a few errors:

- Pylint: my fault, I was using 4 spaces. Fixing.
- Scipy import error: for some CI builds, the scikit-learn version was pinned to a 2016 version that now has some broken functionality that these new changes were relying on. I bumped the version.
- Python 2 errors: I'm ignoring these.
- API changes: it's complaining about changes to internal parameters that should never have been public in the first place. I now appended an `_` to them to fix this for the future. It also doesn't like that I capitalized `X` to match scikit-learn.
- Windows builds: I have no idea why these are failing. Any advice would be much appreciated.

**Comment by k-w-w** [comment]
Thanks for fixing the bugs! I'm pretty sure the windows tests are unrelated. Running the tests again

File: tensorflow/python/keras/wrappers/scikit_learn.py
Comment (k-w-w) [inline tensorflow/python/keras/wrappers/scikit_learn.py]: (Cause of presubmit errors)

We still support python 2, so can you rewrite this?

Comment (adriangb) [reply tensorflow/python/keras/wrappers/scikit_learn.py]: Thanks for looking into this. I've been trying to run the tests locally, but was having issues... I thought tf only supports >= 3.5?

Comment (k-w-w) [reply tensorflow/python/keras/wrappers/scikit_learn.py]: While yes, external Tensorflow requires >= 3.5, Tensorflow is used by google-internal code, some of which is still running on Python 2.

Comment (adriangb) [reply tensorflow/python/keras/wrappers/scikit_learn.py]: I see, thank you for clarifying! I wrote everything for Python 3, but will try to backport and ping here once done.

Comment (adriangb) [reply tensorflow/python/keras/wrappers/scikit_learn.py]: @k-w-w I think we should be good to try CI again

**Comment by k-w-w** [comment]
@fchollet can you review the API changes made to the Scikit Learn wrapper?

**Comment by adriangb** [comment]
@k-w-w fyi the tests only passed just now because I changed the tag in the `BUILD` from `small` to `large` to reflect that they used to take ~10s but with all of the new test cases take ~5m, but I now realize that that filtered them out from CI, so I'm going to change it to `medium` which seems to do the trick. I'm still working on running the CI on my end, compiling takes several hours on my HW

**Comment by adriangb** [comment]
@k-w-w I think everything should be working, passing the tests under the CPU Docker container. I also tested on Python 3.6 & did the sanity check. Could you kick off CI tests again so we can confirm?

**Comment by k-w-w** [comment]
Sure thing, thanks!

**Comment by adriangb** [comment]
Hmm it seems that there are still SciPy/Scikit-Learn version issues. I don't get any issues when running the testing Docker container locally.
Maybe we can bump to `scikit-learn==0.22.2`? 
I wish I could be more methodical but I don't see how I can reproduce the CI runs locally (other than the single test image, which as mentioned runs fine).

File: tensorflow/tools/ci_build/release/common.sh
Comment (adriangb) [inline tensorflow/tools/ci_build/release/common.sh]: @amahendrakar could you please confirm if these are the right scripts to bump scipy/scikit-learn versions for testing? Is there anywhere else I should bump it to ensure there are no issues?
I am adding tests that require these versions as a minimum. Scikit-learn is not required to build, only test, and the scipy version required to build does not need to change.

Comment (adriangb) [reply tensorflow/tools/ci_build/release/common.sh]: @k-w-w if you are able to ping someone internally to verify this that would be very helpful, I don't want to keep asking you to kick off the tests blindly, also don't want to make unnecessary changes or find out that things break under a deployment that CI does not test for. Thanks!

Comment (k-w-w) [reply tensorflow/tools/ci_build/release/common.sh]: @gunan Do you know if this is the right approach to change the scikit versions?

Comment (adriangb) [reply tensorflow/tools/ci_build/release/common.sh]: I'll add that I was able to reproduce the issue on my end, it definitely stems from having a newer version of scipy and an older version of scikit-learn. I had assumed that since the version was not pinned here, it was pulling the latest. But now I'm thinking maybe it has an older version cached or already installed in that image so it's just not bothering to update it. 

It would be nice if CI logged a `pip freeze`, that would help sort out problems like this in the future.

Comment (gunan) [reply tensorflow/tools/ci_build/release/common.sh]: Unfortunately, the CI uses remote build execution, which means we internally need to update some docker images, and push new containers. @av8ramit should be able to point you to instructions for that.

Comment (av8ramit) [reply tensorflow/tools/ci_build/release/common.sh]: Here are our docker containers for CPU/GPU presubmits.
[CPU](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/ci_build/Dockerfile.rbe.ubuntu16.04-manylinux2010)
[GPU](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/ci_build/Dockerfile.rbe.cuda10.1-cudnn7-ubuntu16.04-manylinux2010)

They invoke [this script](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/ci_build/install/install_pip_packages.sh) for installing pypi images.

Unfortunately, someone internally needs to update the containers after this change. @k-w-w I can send you a doc internally with how to update the images.

Comment (adriangb) [reply tensorflow/tools/ci_build/release/common.sh]: @av8ramit I can't see the results from the last CI run anymore because I pushed a new commit, but I do remember the tests passed on the Mac/Python2 build. Is there any way to run again and get the package versions of scipy and scikit-learn that each presubmit runs with to at least verify the issue before making any changes to the images?

**Comment by adriangb** [comment]
Thank you for re-running @av8ramit. I saved the logs this time.

Just to confirm the issue, the [failed test on `Ubuntu CPU`](https://source.cloud.google.com/results/invocations/a35d40fa-4cc0-4161-a226-77a1dfdb9fa8/targets/%2F%2Ftensorflow%2Fpython%2Fkeras%2Fwrappers:scikit_learn_test/log) fails on a scikit-learn line that was changed in this [2017 commit](https://github.com/scikit-learn/scikit-learn/commit/95aa2952e1544a5a2e0f14d366bae1bfd8e9195a).

Because the test passed on [MacOS Python2](https://source.cloud.google.com/results/invocations/01679038-c190-4230-a4f8-478653f0bd39/targets;query=wrapper;collapsed=/%2F%2Ftensorflow%2Fpython%2Fkeras%2Fwrappers:scikit_learn_test/tests;failed=true;notrun=true) and [MacOS Python3](https://source.cloud.google.com/results/invocations/b6b24f79-efb9-4238-871e-e0f724eff514/targets;query=wrapper;collapsed=/%2F%2Ftensorflow%2Fpython%2Fkeras%2Fwrappers:scikit_learn_test), I think the root cause of the failure on `Ubuntu CPU` is indeed an old version of scikit-learn.

If you or @k-w-w can fix this internally that would be much appreciated.

**Comment by av8ramit** [comment]
Hey @adriangb we discussed offline and we are going to wait for general approval for the change and then we can rebuild the containers with the requested changes.

**Comment by adriangb** [comment]
That sounds good to me. I think the main review needed is from @fchollet right?

File: tensorflow/python/keras/wrappers/scikit_learn.py
Comment (fchollet) [inline tensorflow/python/keras/wrappers/scikit_learn.py:233]: `predict_classses` is now deprecated.

**Review by fchollet — COMMENTED** [review]
Hello Adrian,

Thank you for the PR. In general, the scikit-learn API wrapper for Keras is not well-maintained at this time. Because we don't have the resources to maintain it (or even to review a very involved PR like yours), we are considering deprecating it.

Since you are a user of this feature and you've already spent a lot of time developing improvements, I would recommend starting a new repository & pip package hosting an up-to-date version of the API wrapper (effectively, do a fork). We could then redirect users of tf.keras.wrappers.scikit_learn to your repository & pip package, while we deprecate this functionality in tf.keras.

What do you think?

**Comment by adriangb** [comment]
I think that could work really well! But I'd like to list out pros and cons to consider

Pros:
- More flexibility. `scikit-learn` can become required for the wrapper, which eliminates a lot of code duplication for this wrapper and removes the need for `scikit-learn` from TF CI.
- Relieves maintenance burden from Keras/TF team.

Cons:
- Improvements in Keras might take a long time to trickle over.
- When issues do crop up in the future, wrapper maintainers (i.e. me) might have to open an issue with Keras/TF to get support in fixing things.

Overall, I think the pros outweigh the cons. 

I will start working on getting a separate repo with CI/publishing working. In the meantime, if you could do a brief review of this PR as it currently stands, that would be super useful to make sure the initial release is as good as possible.

**Comment by adriangb** [comment]
Are there any preferences as far as:
- Naming of the repo/package or any project this should live under. I came up with `sklearn-keras-wrap`
- Documentation style or content. I planned on copying the [existing docs](https://www.tensorflow.org/api_docs/python/tf/keras/wrappers/scikit_learn) but just keeping everying in the README.md for simplicity.
- Python versions: I think it would be easier to make this package >=Python3.5
- Linting: I would switch to flake8/black
- Testing: I would switch to pytest.

I played around with a packaging a bit, everything seems to work as far as CI/testing/releasing. It would have `scikit-learn>=0.21.0` and `tensorflow>=2.1.0` as dependencies.

**Comment by adriangb** [comment]
A quick update: the package is now fully operational. I settled on the name SciKeras.
PyPi: https://pypi.org/project/scikeras/
GitHub: https://github.com/adriangb/scikeras

Some important updates since this PR:
1. Inherited BaseWrapper from sklearn.base.BaseEstimator.
2. Implemented tags interface.
3. Use OneHotEncoder/LabelEncoder instead of manual numpy work.
4. Fix model serialization bugs.
5. A lot of cleanup.

With all of this, estimators created with these wrappers now pass all of scikit-learn's estimator checks, except those that require setting a random state. As far as I understand, it is not possible to easily set a random seed in tf.

**Comment by adriangb** [comment]
Hi @gbaned, just checking if there are any updates on this proposal/PR? Thanks!

**Comment by gbaned** [comment]
> Hi @gbaned, just checking if there are any updates on this proposal/PR? Thanks!

Hi @adriangb, It is waiting for approval. Thanks!

**Comment by fchollet** [comment]
Thanks for the update. It's great to see that you've already released the new package. We can recommend that people start using it instead of `keras.wrappers.scikit_learn`.

> With all of this, estimators created with these wrappers now pass all of scikit-learn's estimator checks, except those that require setting a random state. As far as I understand, it is not possible to easily set a random seed in tf.

This should be fixable: https://www.tensorflow.org/api_docs/python/tf/random/set_seed

What do you want us to do with the current PR? Should we close it?

**Comment by adriangb** [comment]
> We can recommend that people start using it instead of `keras.wrappers.scikit_learn`.

That sounds good.

> This should be fixable: https://www.tensorflow.org/api_docs/python/tf/random/set_seed

Will take a look, thank you.

> What do you want us to do with the current PR? Should we close it?

I think let's keep it open for a bit longer. Dask is looking to adopt SciKeras as a wrapper ([here](https://github.com/dask/dask-ml/issues/696)), so as they do their testing I expect there to be a couple of issues that crop up in the next couple of weeks that I may need input from the TF team on. Unless the TF team is willing to check the SciKeras repo if they are tagged.

**Comment by fchollet** [comment]
> I think let's keep it open for a bit longer. Dask is looking to adopt SciKeras as a wrapper (here), so as they do their testing I expect there to be a couple of issues that crop up in the next couple of weeks that I may need input from the TF team on. Unless the TF team is willing to check the SciKeras repo if they are tagged.

Ok, sounds good! Please reach out if you need anything from us (over email preferably, so we don't miss it). We'll start recommending your library as soon as it starts getting tractions then 👍

**Comment by gbaned** [comment]
@adriangb  Any update on this PR? Please. Thanks!

**Comment by gbaned** [comment]
@adriangb Any update on this PR? Please. Thanks!

**Comment by adriangb** [comment]
Hi @gbaned, as per François' comment above, the plan is to _not_ merge this PR and instead move this part of tf.keras to an external package. I had asked to keep this PR open for communication and help with bringup of the external package, but I've since established communication with François directly and the external package is making good progress, so I think we can close this PR 😄

**Pull request closed without merging**
